### ISSUE TITLE
DeclarePublicAPIAnalyzer.ReadPublicSymbols seems to be consuming a lot of CPU

### ISSUE THREAD

**Issue by amcasey** [dotnet/roslyn#610]
I installed the newest signed build (17.2) and noticed that VS felt sluggish.  Here's the hot path from the profiler:

Function Name | Inclusive Samples | Exclusive Samples | Inclusive Samples % | Exclusive Samples % | Module Name
------------- | ----------------- | ----------------- | ------------------- | ------------------- | -----------
devenv.exe | 39,010 | 0 | 100.00 | 0.00 | 
>Roslyn.Diagnostics.Analyzers.ApiDesign.DeclarePublicAPIAnalyzer+<>c.<Initialize>b__8_0 | 30,588 | 1 | 78.41 | 0.00 | Roslyn.Diagnostics.Analyzers
>>Roslyn.Diagnostics.Analyzers.ApiDesign.DeclarePublicAPIAnalyzer.ReadPublicSymbols | 30,385 | 481 | **77.89** | 1.23 | Roslyn.Diagnostics.Analyzers
>>>Microsoft.CodeAnalysis.Text.Extensions+SnapshotSourceText+LineInfo.get_Item | 15,231 | 15,231 | 39.04 | 39.04 | Microsoft.CodeAnalysis.EditorFeatures.Text.dll
>>>Microsoft.CodeAnalysis.Text.Extensions+SnapshotSourceText.ToString | 9,201 | 9,201 | 23.59 | 23.59 | Microsoft.CodeAnalysis.EditorFeatures.Text.dll

**Comment by amcasey** [comment]
cc @tmeschter

**Comment by amcasey** [comment]
I should mention that I'm working in the portable-pdb branch.  Maybe the analyzer is assuming the presence of a file that only exists in master?

**Comment by mattwar** [comment]
Is it reading them over and over again?

**Comment by tmeschter** [comment]
@amcasey:
The DeclarePublicAPI analyzer bails out almost immediately if the PublicAPI.txt additional file isn't found.
Which project were you primarily working in? CodeAnalysis.csproj?

@mattwar:
Yes, the analyzer reads the PublicAPI.txt additional file every time the CompilationStart action is called.

**Comment by amcasey** [comment]
@tmeschter I was working on the Expression Compiler projects.

**Comment by amcasey** [comment]
@dpoeschl noticed that it was also using quite a lot of threads.

**Comment by tmeschter** [comment]
@amcasey @dpoeschl That doesn't surprise me. If we saturate the existing threadpool threads with long-running Tasks (like we're doing here) then more threads will be spun up to handle the backlog.

**Comment by dpoeschl** [comment]
Yeah, I saw what appeared to be a hang in completion. Investigated with @rchande and saw 87 and then 155 managed threads (I'm guessing one per project?) that were all in the DeclarePublicAPI analyzer, I think all waiting for a call to GetAdditionalDocument. The completion model task just couldn't seem to get a thread to run on during this time. Things eventually settled down and the completion task did run and the completion list showed up.

**Comment by dpoeschl** [comment]
Also, @tmeschter, if it's relevant, I was typing in EditorFeatures.

**Comment by tmeschter** [comment]
@dpoeschl Did you start to see this after installing a recent build of Roslyn like @amcasey?

**Comment by dpoeschl** [comment]
@tmeschter Yeah, Main-Signed-Release 20150219.1

**Comment by tmeschter** [comment]
@mavasani We've been using the current toolset compilers and diagnostics for several weeks, so the problem can't just be with the analyzer itself. Has something in the analyzer driver changed recently that might cause it to call the compilation start action more frequently while editing a file?

**Comment by mavasani** [comment]
@tmeschter You are right, I think this is probably due to a bug I introduced in the recent refactoring of consolidating analyzer Initialize method and CompilationStartAction execution for both IDE and compiler drivers. Previously we used to execute these methods on the calling threads itself, which could be a UI thread. In the new design, we create a task to run both these analyzer methods and cache the registered actions. The bug is we should create a new task only if one hasn't been already kicked off, rest of the threads should re-use this task. I'll work on a fix ASAP.
